### DISCUSSION TITLE
useCookie in a middleware returns undefined value client side, but real value server side

### DISCUSSION THREAD

**Discussion by noook** [nuxt/nuxt#16460] [Questions] [answered]
My JWT is stored in the `token` cookie and I'm trying to setup a middleware like a route guard to redirect users out of `/admin/*` pages. I am using this following piece of code. Is there something I'm missing ? Or should I do differently ?

I don't know if this is relevant, but I'm inside a layout. It appeared the middleware won't work on the layout, so it's on the `/admin` root page.

```ts
export const useToken = () => useCookie('token', {
  httpOnly: true,
  secure: true,
  sameSite: true,
  path: '/',
});

export default defineNuxtRouteMiddleware(() => {
  const token = useToken();

  console.log(token.value); // undefined in browser, good JWT on server

  if (!token.value) {
    return navigateTo('/');
  }

  const user = decodeJwt(token.value);

  if (!user.roles.includes(ROLE_ADMIN)) {
    return navigateTo('/');
  }
});
```

**Comment by Diizzayy** [comment, marked as answer]
@noook The behavior your experiencing could very well be due to the `httpOnly` flag on your cookie which prevents it from being read client-side. Hence, you'll either have to forego using that flag ( _I wouldn't recommend this_ ) or store the cookie in your app's context on server side initialization for use client side..

If this is indeed the case, I'd recommend that you use a server plugin to extract the token from your cookie and perhaps store it with [useState](https://v3.nuxtjs.org/guide/features/state-management). That would allow you to utilize it throughout your app without having to read and decode the cookie unnecessarily numerous times.

`composables/user.ts`
```ts
export const useUser = () => {
  const user = useState(() => null)
  const token = useCookie('token', {
    httpOnly: true,
    secure: true,
    sameSite: true,
    path: '/'
  })
}
```

`plugins/auth.server.ts`
```ts
export default defineNuxtPlugin(() => {
  const { user, token } = useUser()

  if (token.value) {
    user.value = decodeJwt(token.value);
  }
})
```

`middleware/admin.ts`
```ts
export default defineNuxtRouteMiddleware(() => {
  const { user } = useUser()

  if (!user.value || !user.value.roles?.includes(ROLE_ADMIN)) {
    return navigateTo('/');
  }
});
```

**Reply by noook** [reply]
I tried this way, it worked like a charm, thank you !

**Reply by Diizzayy** [reply]
@noook No problem at all, Happy to hear it worked for yo!

**Reply by noook** [reply]
sorry to bother you, actually I really need the cookie value (which is a JWT) so I can communicate with an external app. So the thing is I have the user on both client and server side however the cookie (containing the JWT) is available only server-side. Do you suggest attaching the JWT to the user object, or do you have another solution ?
